.github/workflows/check.yml

```diff
@@ -1,53 +1,82 @@
-name: check
+name: Check
 
-on: [pull_request, push]
+on:
+  push:
+  pull_request:
+    branches:
+      - master
 
 jobs:
+  check_duplicate_runs:
+    name: Check for duplicate runs
+    continue-on-error: true
+    runs-on: ubuntu-latest
+    outputs:
+      should_skip: ${{ steps.skip_check.outputs.should_skip }}
+    steps:
+      - id: skip_check
+        uses: fkirc/skip-duplicate-actions@master
+        with:
+          concurrent_skipping: always
+          cancel_others: true
+          skip_after_successful_duplicate: true
+          paths_ignore: '["**/README.md", "**/CHANGELOG.md", "**/LICENSE.md"]'
+          do_not_skip: '["pull_request"]'
+
   mix_check:
     name: mix check (Elixir ${{matrix.elixir}}, OTP ${{matrix.otp}})
+    runs-on: ubuntu-18.04
+    needs: check_duplicate_runs
+    if: ${{ needs.check_duplicate_runs.outputs.should_skip != 'true' }}
 
     strategy:
       matrix:
-        include:
-          - elixir: '1.10.4'
-            otp: '23.0.2'
-          - elixir: '1.7.0'
-            otp: '19.0'
-
-    runs-on: ubuntu-18.04
+        elixir:
+        - "1.7"
+        - "1.8"
+        - "1.9"
+        - "1.10"
+        - "1.11"
+        otp:
+        - "20.0"
+        - "21.0"
+        - "22.0.2"
+        - "23.0"
+        exclude:
+        - elixir: "1.10"
+          otp: "20.0"
+        - elixir: "1.11"
+          otp: "20.0"
 
     steps:
-      - name: Checkout
-        uses: actions/checkout@v2
+    - name: Checkout
+      uses: actions/checkout@v2
 
-      - name: Setup Elixir
-        uses: actions/setup-elixir@v1
-        with:
-          elixir-version: ${{ matrix.elixir }}
-          otp-version: ${{ matrix.otp }}
+    - name: Set up Elixir
+      uses: erlef/setup-elixir@v1
+      with:
+        elixir-version: ${{ matrix.elixir }}
+        otp-version: ${{ matrix.otp }}
 
-      - name: Cache Elixir deps
-        uses: actions/cache@v2
-        with:
-          path: deps
-          key: elixir-deps-${{ runner.os }}-${{ hashFiles('mix.lock') }}
-          restore-keys: |
-            elixir-deps-${{ runner.os }}-
+    - name: Restore deps cache
```

Comment:
So your proposition is to cache both deps and build in one step... Not sure if that's actually an improvement since that seems to create the problem that you've described i.e. having to add commit sha at the end of cache key in order to store up-to-date build. With previous solution, all files that could've affected the build were included in hashFiles so, assuming the compile task is idempotent, there shouldn't ever be a need to update cache if they didn't change. That solution seems more explicit to me, but OTOH it may work slower if hashFiles takes time for entire lib (not sure if it does). Are there any scenarios in which your solution is clearly superior? You've mentioned dialyzer but ex_check doesn't use it (and even if it would, the assumption about idempotency should still hold right?). What do you think?

I case we'd go with your solution, I'd definitely rename the step to "Restore deps and build cache" or similar as currently it only tells half of the story :)

Reply:
`deps` contains the source code of dependencies. `_build` contains the compiled `.beam` files for these dependencies as well as for your project. And compiled files of the project are usually just the small part of `_build` folder content. So there is no actual reason to store these folders separately, because updating files in `deps` will lead to updating the `_build` folder. What's why I moved both of then into one caching step with the same key.

"and even if it would, the assumption about idempotency should still hold right?"
Well, dialyzer generates .plts files for the project and used dependencies. By default it stores project .plts files in `_build/#{Mix.env()}` folder and deps .plts files in `deps` folder. Some other instruments can manipulate it too.

So all the content of these dirs depends on the hash of all the files in the repo. Calculating this hash does not take so long because even for large files. But you already have a git repo where hashes of all the files are already combined into one commit hash. What's why I'm using it as a cache key.

`actions/cache` step will search for existing cache by this key, and because every commit hash is different, it will never hit. But then it will start to search for recent cache items by prefixes listed in restore-keys, and because of that you'll always have the most recent cache data as possible.

Is something is different in cache and in mix.lock or other files, that's not a problem because Elixir compiles will throw away compiled .beam files if they are outdated. dializer can detect these changes too, and in such a case it will remove old sources from .plts files and add new ones.

Overall impact on the CI run speed is huge - for example, in one of my projects compiling all the dependencies could take about  minutes and generating fresh .plts files for dialyzer can take another 20 minutes. But with cache enabled it'll take only 2 minutes to run all the CI job.